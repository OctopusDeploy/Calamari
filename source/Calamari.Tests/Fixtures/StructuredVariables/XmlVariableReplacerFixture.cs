using System;
using System.Runtime.CompilerServices;
using Assent;
using Calamari.Common.Features.StructuredVariables;
using Calamari.Common.Plumbing.FileSystem;
using Calamari.Common.Plumbing.Variables;
using Calamari.Tests.Helpers;
using NUnit.Framework;

namespace Calamari.Tests.Fixtures.StructuredVariables
{
    [TestFixture]
    public class XmlVariableReplacerFixture : VariableReplacerFixture
    {
        public XmlVariableReplacerFixture()
            : base(new XmlFormatVariableReplacer(CalamariPhysicalFileSystem.GetPhysicalFileSystem(), new InMemoryLog()))
        {
        }

        [Test]
        public void DoesNothingIfThereAreNoVariables()
        {
            var vars = new CalamariVariables();
            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceAComment()
        {
            var vars = new CalamariVariables
            {
                { "/document/comment()", "New Comment" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceAnAttribute()
        {
            var vars = new CalamariVariables
            {
                { "/document/setting[@id='id-1']/@id", "id-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceATextNode()
        {
            var vars = new CalamariVariables
            {
                { "/document/setting[@id='id-1']/text()", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceAnElementsText()
        {
            var vars = new CalamariVariables
            {
                { "/document/setting[@id='id-1']", "value<new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanInsertTextIntoAnEmptyElement()
        {
            var vars = new CalamariVariables
            {
                { "/document/empty", "value<new" }
            };

            RunTest(vars, "elements.xml");
        }

        [Test]
        public void CanInsertTextIntoASelfClosingElement()
        {
            var vars = new CalamariVariables
            {
                { "/document/selfclosing", "value<new" }
            };

            RunTest(vars, "elements.xml");
        }

        [Test]
        public void CanReplaceAnElementsChildrenWhenTheElementHasMixedContent()
        {
            var vars = new CalamariVariables
            {
                { "/document/mixed", "<newElement />" }
            };

            RunTest(vars, "elements.xml");
        }

        [Test]
        public void CanReplaceAnElementsChildren()
        {
            var vars = new CalamariVariables
            {
                { "//moreSettings", "<a /><b />" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceAnElementsChildrenWhenTheNewChildrenHaveNamespaces()
        {
            var vars = new CalamariVariables
            {
                { "//moreSettings", "<db:a />" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void DoesNotModifyDocumentWhenVariableCannotBeParsedAsMarkup()
        {
            var vars = new CalamariVariables
            {
                { "//moreSettings", "<<<<" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void DoesntTreatVariableAsMarkupWhenReplacingAnElementThatContainsNoElementChildren()
        {
            var vars = new CalamariVariables
            {
                { "/document/setting[@id='id-1']", "<a />" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceMultipleElements()
        {
            var vars = new CalamariVariables
            {
                { "/document/setting", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceMultipleElementsInDifferentPartsOfTheTree()
        {
            var vars = new CalamariVariables
            {
                { "//setting", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceCData()
        {
            var vars = new CalamariVariables
            {
                { "/document/setting[@id='id-3']/text()", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceCDataInAnElement()
        {
            var vars = new CalamariVariables
            {
                { "/document/setting[@id='id-3']", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanReplaceProcessingInstructions()
        {
            var vars = new CalamariVariables
            {
                { "/document/processing-instruction('xml-stylesheet')", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanInferNamespacePrefixesFromDocument()
        {
            var vars = new CalamariVariables
            {
                { "/document/unique:anotherSetting", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void CanUseXPath2Wildcards()
        {
            var vars = new CalamariVariables
            {
                { "//*:node", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void DoesNotThrowOnUnrecognisedNamespace()
        {
            var vars = new CalamariVariables
            {
                { "//env:something", "value-new" }
            };

            RunTest(vars, "complex.xml");
        }

        [Test]
        public void UsesTheFirstNamespaceWhenADuplicatePrefixIsFound()
        {
            var vars = new CalamariVariables
            {
                { "//parent/dupe:node", "value-new" }
            };

            RunTest(vars, "duplicate-prefixes.xml");
        }

        [Test]
        public void ShouldIgnoreOctopusPrefix()
        {
            var vars = new CalamariVariables
            {
                { "Octopus.Release.Id", "999" }
            };
            RunTest(vars, "ignore-octopus.xml");
        }

        [Test]
        public void HandlesVariablesThatReferenceOtherVariables()
        {
            var vars = new CalamariVariables
            {
                { "/document/selfclosing", "#{a}" },
                { "a", "#{b}" },
                { "b", "value-new" },
            };
            RunTest(vars, "elements.xml");
        }

        [Test]
        public void ShouldPreserveEncodingUtf8DosBom()
        {
            this.Assent(ReplaceToHex(new CalamariVariables(), "enc-utf8-dos-bom.xml"),
                        TestEnvironment.AssentConfiguration);
        }

        [Test]
        public void ShouldPreserveEncodingUtf8UnixNoBom()
        {
            this.Assent(ReplaceToHex(new CalamariVariables(), "enc-utf8-unix-nobom.xml"),
                        TestEnvironment.AssentConfiguration);
        }

        [Test]
        public void ShouldPreserveEncodingUtf16DosBom()
        {
            this.Assent(ReplaceToHex(new CalamariVariables(), "enc-utf16-dos-bom.xml"),
                        TestEnvironment.AssentConfiguration);
        }

        [Test]
        public void ShouldPreserveEncodingWindows1252DosNoBom()
        {
            this.Assent(ReplaceToHex(new CalamariVariables(), "enc-windows1252-dos-nobom.xml"),
                        TestEnvironment.AssentConfiguration);
        }

        void RunTest(CalamariVariables vars,
                     string file,
                     [CallerMemberName]
                     string testName = null,
                     [CallerFilePath]
                     string filePath = null)
        {
            // ReSharper disable once ExplicitCallerInfoArgument
            this.Assent(Replace(vars, file), TestEnvironment.AssentXmlConfiguration, testName, filePath);
        }
    }
}