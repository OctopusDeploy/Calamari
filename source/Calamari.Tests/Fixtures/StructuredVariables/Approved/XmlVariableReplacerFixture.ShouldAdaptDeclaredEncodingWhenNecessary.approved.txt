3C 3F 78 6D 6C 20 76 65   <?xml ve
72 73 69 6F 6E 3D 22 31   rsion="1
2E 30 22 20 65 6E 63 6F   .0" enco
64 69 6E 67 3D 22 75 74   ding="ut
66 2D 38 22 3F 3E 0A 3C   f-8"?>^<
74 6F 70 2D 6E 61 6D 65   top-name
73 2D 66 6F 72 2D 62 6F   s-for-bo
79 73 3E 0A 20 20 3C 6E   ys>^  <n
61 6D 65 3E 4A 61 6B 75   ame>Jaku
62 3C 2F 6E 61 6D 65 3E   b</name>
0A 20 20 3C 6E 61 6D 65   ^  <name
3E 53 7A 79 6D 6F 6E 3C   >Szymon<
2F 6E 61 6D 65 3E 0A 20   /name>^ 
20 3C 6E 61 6D 65 3E 4B    <name>K
61 63 70 65 72 3C 2F 6E   acper</n
61 6D 65 3E 0A 20 20 3C   ame>^  <
6E 61 6D 65 3E 46 69 6C   name>Fil
69 70 3C 2F 6E 61 6D 65   ip</name
3E 0A 20 20 3C 6E 61 6D   >^  <nam
65 3E 4D 69 63 68 61 C5   e>Micha?
82 3C 2F 6E 61 6D 65 3E   ?</name>
0A 20 20 3C 6E 61 6D 65   ^  <name
3E 4D 61 74 65 75 73 7A   >Mateusz
3C 2F 6E 61 6D 65 3E 0A   </name>^
20 20 3C 6E 61 6D 65 3E     <name>
42 61 72 74 6F 73 7A 3C   Bartosz<
2F 6E 61 6D 65 3E 0A 20   /name>^ 
20 3C 6E 61 6D 65 3E 57    <name>W
6F 6A 63 69 65 63 68 3C   ojciech<
2F 6E 61 6D 65 3E 0A 20   /name>^ 
20 3C 6E 61 6D 65 3E 41    <name>A
64 61 6D 3C 2F 6E 61 6D   dam</nam
65 3E 0A 20 20 3C 6E 61   e>^  <na
6D 65 3E 57 69 6B 74 6F   me>Wikto
72 3C 2F 6E 61 6D 65 3E   r</name>
0A 20 20 3C 6E 61 6D 65   ^  <name
3E 50 69 6F 74 72 3C 2F   >Piotr</
6E 61 6D 65 3E 0A 20 20   name>^  
3C 6E 61 6D 65 3E 4A 61   <name>Ja
6E 3C 2F 6E 61 6D 65 3E   n</name>
0A 20 20 3C 6E 61 6D 65   ^  <name
3E 44 61 77 69 64 3C 2F   >Dawid</
6E 61 6D 65 3E 0A 20 20   name>^  
3C 6E 61 6D 65 3E E5 BC   <name>??
A0 E4 BC 9F 3C 2F 6E 61   ????</na
6D 65 3E 0A 3C 2F 74 6F   me>^</to
70 2D 6E 61 6D 65 73 2D   p-names-
66 6F 72 2D 62 6F 79 73   for-boys
3E                        >